src/amber.cc

```diff
@@ -86,6 +88,57 @@ amber::Result Amber::ExecuteWithShaderData(const amber::Recipe* recipe,
     return r;
   }
 
+  if (!opts.image_file.empty()) {
+
+    // Write framebuffer image to PPM file
+
+    ResourceInfo framebuffer_info;
+    r = engine->GetFrameBufferInfo(&framebuffer_info);
+    if (!r.IsSuccess()) {
+      return r;
+    }
+
+    if (framebuffer_info.type != ResourceInfoType::kImage) {
+      return Result("Resource is not an image");
+    }
+
+    const uint32_t width = framebuffer_info.image_info.width;
+    const uint32_t height = framebuffer_info.image_info.height;
+
+    FormatBuffer pixels(BufferType::kFramebuffer);
+    pixels.SetName("Framebuffer pixels");
+    r = engine->GetFrameBuffer(&pixels);
+    if (!r.IsSuccess()) {
+      return r;
+    }
+
+    if (pixels.GetFormat().GetFormatType() != FormatType::kR8G8B8A8_UINT) {
+      return Result("Only R8G8B8A8 is supported for now");
+    }
+
+    std::ofstream image_stream;
+    image_stream.open(opts.image_file, std::ios::out | std::ios::binary);
+    if (!image_stream.is_open()) {
+      return Result("Cannot open file to write image");
+    }
+
+    // PPM header
```

Comment:
If this gets pulled out this would then live in samples/ppm.h/cc or something like that

Reply:
In that case, adding a unittest for this would be nice.
We can use a hard-coded image and run this code for the image to check the output PPM is correct.

Reply:
See my new push for this. An issue is: how do you properly define tests for modules under `samples/`? Also, by hard-coded image if you meant a file, where should this file be / how can the test know the path to this file? Otherwise, store the hard-coded image in the test source as an array of bytes?

I tried my best, happy to hear about your feedback on this.